### ISSUE TITLE
Login-field for editors

### ISSUE THREAD

**Issue by drjakob** [tactilenews/100eyes#418]
As editor
I want to have a normal Login-field at the landing page to enter my credentials
in order to feel more convenient.

**Comment by drjakob** [comment]
![Uploading Bildschirmfoto 2020-10-08 um 01.19.56.png…]()

**Comment by drjakob** [comment]
<img width="803" alt="Bildschirmfoto 2020-10-08 um 01 19 56" src="https://user-images.githubusercontent.com/11591426/95397836-858c0f80-0904-11eb-8a5b-2c09f443011e.png">

**Comment by tillprochaska** [comment]
@mattwr18 @phoeinx: Because today we were unsure if this ticket is till high prio: I checked with @drjakob and @FrauCsu and it still is, so if you’re out of work, maybe look into this one.

I think we might want to clarify some of the details and the concrete use case. But as discussed earlier, there are still a few open questions:

* This ticket is about log in to our web app, i.e. login for editors, not contributors.
* I guess the UI doesn’t have to look exactly as pictured above.
* We don’t want a "Neu registrieren" (self sign-up) feature (otherwise, anyone would be able to create and editor account…)
* Even if a UI to manage (create/delete) users would be nice, I think it would be fine to add that in a second step to keep PRs small.
* Same for the password reset feature: I think this may be solved in a second step. (I’m not sure wether this is important right now at all.)  

Implementation-wise, I think we should go for one of the established user auth solutions available for Rails:
* https://github.com/binarylogic/authlogic
* https://github.com/heartcombo/devise

**Comment by mattwr18** [comment]
I looked at at least two different ways to accomplish this task. The first seems the simplest and would involve just having a log in form that would check that password matches the single user created before the campaign. The second way, which I pushed up the code for is to add devise to the project.

The first, I believe, is simpler, but has limitations in that it does not support things like password resets.
The second came with some unexpected issues since Devise validates that a user is created with a password. See the build, but here are `126/273` failures.

I found an issue on [SO of the Devise validations](https://stackoverflow.com/questions/32185562/devise-create-users-without-password), which gives several potential workarounds, one of which is the gem [devise_invitable](https://github.com/scambra/devise_invitable) 

Also, we discussed maybe having a separate model for `Contributors`, which would not then have devise on the model since we don't require a password. I guess this might be quite a bit of work  for this `PR`...

Question to DEVs, @tillprochaska @roschaefer @phoeinx, what do you think is the best way forward? Should we have a simple as possible solution for now, or should I go forward with trying to get Devise working?

**Comment by tillprochaska** [comment]
Hi @mattwr18, here are my thoughts on this.

> The first seems the simplest and would involve just having a log in form that would check that password matches the single user created before the campaign.

When we set up the current login solution (using a single pair of login credentials and HTTP Basic Auth), it was always clear that we would have to implement real user management at some point.

You’re right, we don’t need most of the features Devise provides right now. But as soon as the software’s used by users other than Astrid, Jakob, and Isa, we’ll probably need them. Also, we’ve already discussed adding 2FA/MFA quite a few times etc.

More importantly, I feel like there are way too many pitfalls in building your own auth solution. Right now, the scope of that solution might be limited, but it will grow, and I’m pretty sure it will cause problems at some point. Therefore, I’d really like to build on an established, robust library like Devise or Authlogic.

> I found an issue on SO of the Devise validations, which gives several potential workarounds, one of which is the gem devise_invitable

I had a quick look at the gem you linked, which says it provides functionality to send invites via email, but eventually users will still have to set a password. Is that really solving the problem?

> Also, we discussed maybe having a separate model for Contributors, which would not then have devise on the model since we don't require a password.

| **Contributors** | **Users** |
| - | - |
| They participate in a project. Except for the onboarding process, they never interact with out application directly. | They use 100eyes (i.e. they need to log in to 100eyes) to send requests, interact with contributors, and manage project settings. |
| Contributors are right now represented by the `User` model (and that’s confusing). | Users don’t exist as a concept in our application yet. |
| Every contributor has the same, limited permissions. | For now, all users have the same permissions. This might change in the future, e.g. certain users might be allowed to managers, while other would only be able to send requests. |
| We store several different attributes for each contributor, including telegram ids, email addresses, contact data, notes. | For now, we’d only store basic login data and possibly a name for each user. If we need to store additional data for users in the future, it’s likely going to be different attributes than the ones we’re storing for contributors. |
| Contributors can be senders and recipients of messages. | Currently, we don’t store which individual user has sent or received a certain message, but that might change at some point. |
| Contributors use the onboarding flow to create an account. | User can’t set up an account themselves. Accounts are set up manually, either by a developer via the Rails console, or – in the future – using a user management UI. |

While there are some similarities (both can send and receive messages) between contributors and users, they are different in many ways.

> I guess this might be quite a bit of work for this PR...

What we probably have to do is migrate away from using the term "user" for people who actually aren’t users of our software, but contributors. While that would certainly be quite some work (we’d have to replace the term "user" in the whole codebase), I could imagine that there is little potential for errors (as it’s basically just changing any occurrence of `user` to `contributor`). And we’d have to add a migration to rename the `users` table to `contributors`.

Adding Devise for contributors, on the other hand, might cause more difficult problems. How do we solve the required email thing? How do we integrate our new onboarding flow with it? 

I’m not 100% sure about this, but I still feel that having them as separate models might be the cleanest solution. While we could handle some of the differences, I’m more afraid that the conventions Devise assumes for user accounts don’t work well with the way contributor accounts work in 100eyes.

I’d really like to hear @phoeinx and @roschaefer on this as well!

**Comment by tillprochaska** [comment]
@mattwr18 Maybe we could split the refactoring from "users" to "contributors" in a separate issue, then add Devise after that’s done? That might reduce the scope of the individual task and the potential for merge conflicts etc. Also, you certainly wouldn’t need to do all of it, if you don’t want to. (I could imagine the refactoring part to be rather boring…)

**Comment by mattwr18** [comment]
I think very similarly to you on this topic, which is great! I prefer the cleaner solution that will save us time and effort down the road, over a temp quick fix with predictable issues down the road. As I said, I just wanted to make this decision together with others, especially those who have a higher level view of the project as a whole.

I wouldn't mind seeing it through from splitting users and contributors, to adding a user management tool. It might be a bit boring, but it will help me to touch many parts of the code base, and the more I familiarize myself, the better for me and the project. I was recently looking into https://github.com/thoughtbot/clearance and considering as an alternative to the two tools you proposed. It seems fairly well maintained, is brought to us by our friends at thoughtbot. If you compare it to Devise, it's not as active or popular https://ruby.libhunt.com/compare-clearance-vs-devise , but neither is Authlogic .... 
One thing to consider is whether we anticipate support OAuth or not.

**Comment by tillprochaska** [comment]
Good to hear that, @mattwr18! 

> I was recently looking into https://github.com/thoughtbot/clearance and considering as an alternative to the two tools you proposed. It seems fairly well maintained, is brought to us by our friends at thoughtbot.

I’d be open to that if you feel it has advantages over Devise, and I like thoughtbot. OAuth *might* come in handy, but I’d imagine that’s nothing we will integrate in the midterm future. I’ve read a comment by the maintainers that they recommend integrating OAuth via the `omniauth` gem and that it should play nicely with `clearance`.

What we should definitely consider is adding 2FA/one time passwords, as that’s something we might need in the short-term future. From what I’ve seen though, it might be fairly easy to implement the basic logic for that ourselves, while relying on the `rotp` gem for the crypto part.

**Comment by tillprochaska** [comment]
> it might be fairly easy to implement the basic logic for that ourselves, while relying on the rotp gem for the crypto part.

On the other hand, if there’s a solution for that using Devise that already works…

**Comment by mattwr18** [comment]
New issue created https://github.com/tactilenews/100eyes/issues/469
I'll put this on hold for now as it's blocked by that issue
